pkg/controller/volume/persistentvolume/scheduler_binder_cache.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package persistentvolume
+
+import (
+	"github.com/golang/glog"
+
+	"sync"
+
+	"k8s.io/api/core/v1"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+// PodBindingCache stores PV binding decisions per pod per node.
+// Pod entries are removed when the Pod is deleted or updated to
+// no longer be schedulable.
+type PodBindingCache interface {
+	// UpdateBindings will update the cache with the given bindings for the
+	// pod and node.
+	UpdateBindings(pod, node string, bindings []*bindingInfo)
+
+	// GetBindings will return the cached bindings for the given pod and node.
+	GetBindings(pod, node string) []*bindingInfo
+}
+
+type podBindingCache struct {
+	mutex sync.Mutex
+
+	// Key = pod name
+	// Value = nodeBindings
+	bindings map[string]nodeBindings
+}
+
+// Key = nodeName
+// Value = array of bindingInfo
+type nodeBindings map[string][]*bindingInfo
+
+func NewPodBindingCache(podInformer coreinformers.PodInformer) PodBindingCache {
+	c := &podBindingCache{bindings: map[string]nodeBindings{}}
+
+	podInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: c.updatePod,
+			DeleteFunc: c.deletePod,
+		},
+	)
+
+	return c
+}
+
+func isPodSchedulable(pod *v1.Pod) bool {
+	if len(pod.Spec.NodeName) != 0 {
+		return false
+	}
+	return pod.Status.Phase == v1.PodPending
```

Comment:
@bsalamat I noticed that in the scheduler, it only disallows pod phases "succeeded" and "failed".  When would you have a pod with no node name, but is in "running" or "unknown" state, and you would want the scheduler to schedule it?

Reply:
Discussed with @bsalamat.  I'll hook this cache into the scheduler's pod informer handlers + scheduling error handlers instead of doing my own thing.